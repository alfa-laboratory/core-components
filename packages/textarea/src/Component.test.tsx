import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Textarea } from './index';

describe('Textarea', () => {
    describe('Snapshots tests', () => {
        it('should match snapshot', () => {
            const { container } = render(<Textarea value='value' />);

            expect(container).toMatchSnapshot();
        });
    });

    it('should forward ref to textarea', () => {
        const textareaRef = jest.fn();
        const dataTestId = 'test-id';
        const { getByTestId } = render(<Textarea ref={textareaRef} dataTestId={dataTestId} />);

        expect(textareaRef.mock.calls).toEqual([[getByTestId(dataTestId)]]);
    });

    it('should set `data-test-id` attribute to textarea', () => {
        const dataTestId = 'test-id';
        const { getByTestId } = render(<Textarea block={true} dataTestId={dataTestId} />);

        expect(getByTestId(dataTestId).tagName).toBe('TEXTAREA');
    });

    describe('Classes tests', () => {
        it('should set `className` class to form-control inner', () => {
            const { container } = render(<Textarea className='test-class' />);

            const inner = container.querySelector('.test-class');

            expect(inner).toHaveClass('inner');
        });

        it('should set `textareaClassName` class to textarea', () => {
            const dataTestId = 'test-id';
            const className = 'test-class';
            const { getByTestId } = render(
                <Textarea textareaClassName={className} dataTestId={dataTestId} />,
            );

            expect(getByTestId(dataTestId)).toHaveClass(className);
        });

        describe('when component is controlled', () => {
            it('should set `filled` class when value passed', () => {
                const dataTestId = 'test-id';
                const { getByTestId } = render(
                    <Textarea value='some value' dataTestId={dataTestId} />,
                );

                expect(getByTestId(dataTestId)).toHaveClass('filled');
            });

            it('should not set `filled` class if the value is empty', () => {
                const dataTestId = 'test-id';
                const { getByTestId } = render(<Textarea value='' dataTestId={dataTestId} />);

                expect(getByTestId(dataTestId)).not.toHaveClass('filled');
            });

            it('should unset `filled` class if the value becomes empty', () => {
                const dataTestId = 'test-id';
                const { getByTestId, rerender } = render(
                    <Textarea value='some value' dataTestId={dataTestId} />,
                );

                rerender(<Textarea value='' dataTestId={dataTestId} />);

                expect(getByTestId(dataTestId)).not.toHaveClass('filled');
            });
        });

        describe('when component is uncontrolled', () => {
            it('should set `filled` class when defaultValue passed', () => {
                const dataTestId = 'test-id';
                const { getByTestId } = render(
                    <Textarea defaultValue='some value' dataTestId={dataTestId} />,
                );

                expect(getByTestId(dataTestId)).toHaveClass('filled');
            });

            it('should not set `filled` class if the value is empty', () => {
                const dataTestId = 'test-id';
                const { getByTestId } = render(<Textarea dataTestId={dataTestId} />);

                expect(getByTestId(dataTestId)).not.toHaveClass('filled');
            });

            it('should unset `filled` class if value becomes empty', async () => {
                const dataTestId = 'test-id';
                const { getByTestId } = render(
                    <Textarea defaultValue='some value' dataTestId={dataTestId} />,
                );

                const textarea = getByTestId(dataTestId) as HTMLInputElement;

                textarea.setSelectionRange(0, textarea.value.length);
                await userEvent.type(textarea, '{backspace}');

                textarea.blur();

                expect(textarea.value).toBe('');
                expect(textarea).not.toHaveClass('filled');
            });
        });

        it('should set `hasLabel` class', () => {
            const dataTestId = 'test-id';
            const { getByTestId } = render(<Textarea label='label' dataTestId={dataTestId} />);

            expect(getByTestId(dataTestId)).toHaveClass('hasLabel');
        });

        it('should set `resizeVertical` class', () => {
            const dataTestId = 'test-id';
            const { getByTestId } = render(<Textarea resize='vertical' dataTestId={dataTestId} />);

            expect(getByTestId(dataTestId)).toHaveClass('resizeVertical');
        });

        it('should set `disabled` attribute', () => {
            const dataTestId = 'test-id';
            const { getByTestId } = render(<Textarea disabled={true} dataTestId={dataTestId} />);

            expect(getByTestId(dataTestId)).toHaveAttribute('disabled');
        });

        it('should set `maxHeight` style with autosize on', () => {
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <Textarea autosize={true} maxHeight={100} dataTestId={dataTestId} />,
            );

            expect(getByTestId(dataTestId)).toHaveStyle('max-height: 100px');
        });

        it('should set `maxHeight` style with autosize off', () => {
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <Textarea autosize={false} maxHeight={100} dataTestId={dataTestId} />,
            );

            expect(getByTestId(dataTestId)).toHaveStyle('max-height: 100px');
        });
    });

    describe('Callbacks tests', () => {
        it('should call `onChange` prop', () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const value = '123';
            const { getByTestId } = render(<Textarea onChange={cb} dataTestId={dataTestId} />);

            const textarea = getByTestId(dataTestId) as HTMLTextAreaElement;

            fireEvent.change(textarea, { target: { value } });

            expect(cb).toBeCalledTimes(1);
            expect(textarea.value).toBe(value);
        });

        it('should call `onFocus` prop', () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(<Textarea onFocus={cb} dataTestId={dataTestId} />);

            fireEvent.focus(getByTestId(dataTestId));

            expect(cb).toBeCalledTimes(1);
        });

        it('should call `onBlur` prop', () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(<Textarea onBlur={cb} dataTestId={dataTestId} />);

            fireEvent.blur(getByTestId(dataTestId));

            expect(cb).toBeCalledTimes(1);
        });

        it('should not call `onChange` prop if disabled', async () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <Textarea onChange={cb} dataTestId={dataTestId} disabled={true} />,
            );

            const textarea = getByTestId(dataTestId) as HTMLTextAreaElement;

            await userEvent.type(textarea, '123');

            expect(cb).not.toBeCalled();
        });
    });

    it('should unmount without errors', () => {
        const { unmount } = render(<Textarea value='value' onChange={jest.fn()} />);

        expect(unmount).not.toThrowError();
    });
});
