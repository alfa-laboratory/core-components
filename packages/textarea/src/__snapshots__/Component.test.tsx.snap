// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea Snapshots tests should match snapshot 1`] = `
<div>
  <div
    class="component s"
  >
    <div
      class="inner filled"
    >
      <div
        class="inputWrapper"
      >
        <div
          class="input"
        >
          <textarea
            autocomplete="on"
            class="textarea s filled"
            rows="1"
          >
            value
          </textarea>
        </div>
      </div>
    </div>
    
  </div>
</div>
`;
